### PR TITLE
Enable AdoptedResource using latest code-gen logic for ScalableTarget

### PR DESCRIPTION
**Pull request by mbaijal**
First two commits are the same as #42 but created a separate PR because I wanted preserve the changes of that PR. 

TODO:
 - I still need to address [one review comment from #42 ](https://github.com/aws-controllers-k8s/applicationautoscaling-controller/pull/42#discussion_r706532786)
 - Runtime update to 0.14.0

Description of changes:
 - Enable AdoptedResource for both resources
 - Add required custom logic for ScalableTarget 
 - Add integration test for the same

Testing
- Manually verified that scalableTarget can be adopted
- Local Kind tests pass 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/e2e/tests/test_adopted_resource.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        assert policy_resource["spec"].get("resourceID", None) == sdk_resource_id
+
+        # Delete the Adopted Resources
+        _, deleted = k8s.delete_custom_resource(adopted_policy_reference)
+        assert deleted is True
+
+        _, deleted = k8s.delete_custom_resource(adopted_target_reference)
+        assert deleted is True
+
+        # Delete the ApplicationAutoscaling Resources
```
Comment (mbaijal) [inline test/e2e/tests/test_adopted_resource.py:196]: @surajkota This is what was missing in the previous PR which meant that boto3 deletion was being used. 

[reference](https://github.com/aws-controllers-k8s/applicationautoscaling-controller/pull/42#discussion_r706532786)

File: pkg/resource/scalable_target/custom_api.go [hunk 2]
```diff
@@ -40,3 +40,13 @@ func (rm *resourceManager) customSetLastModifiedTime(ko *svcapitypes.ScalableTar
 	currentTime := metav1.Time{Time: time.Now().UTC()}
 	ko.Status.LastModifiedTime = &currentTime
 }
+
+// customCheckRequiredFieldsMissingMethod returns true if there are any fields
+// for the ReadOne Input shape that are required but not present in the
+// resource's Spec or Status
+func (rm *resourceManager) customCheckRequiredFieldsMissingMethod(
```
Comment (RedbackThomson) [inline pkg/resource/scalable_target/custom_api.go:47]: nit: Remove `Method` from the method name. We know it's a method :)

File: pkg/resource/scalable_target/resource.go [hunk 3]
```diff
@@ -95,6 +95,7 @@ func (r *resource) SetIdentifiers(identifier *ackv1alpha1.AWSIdentifiers) error
 	if identifier.NameOrID == "" {
 		return ackerrors.MissingNameIdentifier
 	}
+	r.ko.Spec.ResourceID = &identifier.NameOrID
```
Comment (RedbackThomson) [inline pkg/resource/scalable_target/resource.go:98]: Yay. This convinced me your `generator.yaml` changes are correct.

File: test/e2e/common/application_autoscaling_utils.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
```
Comment (RedbackThomson) [inline test/e2e/common/application_autoscaling_utils.py:1]: nit: Change this file name to `utils.py`. I saw the conversation on the previous PR about this, but I agree with Suraj that the default for this repository is application autoscaling, so no need to preface anything with that.

File: test/e2e/tests/test_adopted_resource.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    yield (resource_id, adopted_target_reference, adopted_policy_reference)
+
+    if k8s.get_resource_exists(adopted_policy_reference):
+        _, deleted = k8s.delete_custom_resource(adopted_policy_reference)
+        assert deleted
+
+
+@service_marker
+@pytest.mark.canary
+class TestAdoptedSageMakerEndpointAutoscaling:
```
Comment (RedbackThomson) [inline test/e2e/tests/test_adopted_resource.py:145]: Please rename to `TestAdoption` and rename the test below to `test_sagemaker_endpoint_autoscaling`. In the future we may add adoption tests for other resources.

File: test/e2e/tests/test_adopted_resource.py (hunk 1 above)
Comment (RedbackThomson) [reply test/e2e/tests/test_adopted_resource.py:196]: Are these necessary? We are testing for deletion in the fixtures (they delete and assert)

**Review by RedbackThomson — CHANGES_REQUESTED** [review]
Thanks for updating this! Tests look good!

File: test/e2e/tests/test_adopted_resource.py (hunk 1 above)
Comment (surajkota) [reply test/e2e/tests/test_adopted_resource.py:196]: yes, delete is getting tested but the intention is also to clean up the resources created by the test. if we don't have this, the actual resource will get deleted and the controller will error if it reconciles

File: pkg/resource/scalable_target/custom_api.go (hunk 2 above)
Comment (mbaijal) [reply pkg/resource/scalable_target/custom_api.go:47]: done

File: test/e2e/tests/test_adopted_resource.py (hunk 1 above)
Comment (mbaijal) [reply test/e2e/tests/test_adopted_resource.py:196]: added delete logic.

File: test/e2e/common/application_autoscaling_utils.py (hunk 4 above)
Comment (mbaijal) [reply test/e2e/common/application_autoscaling_utils.py:1]: done

File: test/e2e/tests/test_adopted_resource.py (hunk 5 above)
Comment (mbaijal) [reply test/e2e/tests/test_adopted_resource.py:145]: done.

File: test/e2e/common/utils.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        "ServiceNamespace": "sagemaker",
+        "ResourceId": resource_id,
+        "ScalableDimension": "sagemaker:variant:DesiredInstanceCount",
+        "PolicyName": policy_name,
+    }
+
+    describe_response = sagemaker_endpoint_describe_scaling_policy(
+        resource_id, policy_name
+    )
+    if len(describe_response["ScalingPolicies"]) > 0:
```
Comment (mbaijal) [inline test/e2e/common/utils.py:98]: can remove `>0`

File: test/e2e/tests/test_adopted_resource.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    sagemaker_make_endpoint,
+)
+from e2e.common.utils import (
+    sagemaker_endpoint_register_scalable_target,
+    sagemaker_endpoint_put_scaling_policy,
+    sagemaker_endpoint_deregister_scalable_target,
+    sagemaker_endpoint_delete_scaling_policy,
+)
+
+TARGET_RESOURCE_PLURAL = "scalabletargets"
```
Comment (surajkota) [inline test/e2e/tests/test_adopted_resource.py:39]: nit: these can move to common config

**Review by surajkota — APPROVED** [review]
/approve

**Pull request merged**